### PR TITLE
Fix gradient clipping on XLA device

### PR DESCRIPTION
**Pull request by ymwangg**
# What does this PR do?
This PR fixed the gradient clipping logic on XLA device. We found all_reduce is wrongfully disabled in fp16 mode when `max_grad_norm=0` on XLA gpu device. To be consistent with native pytorch behavior, all_reduce should be placed immediately after calling `self.training_step`.

Tested with the run_mlm.py example script on 8 Nvidia V100 GPU:
```sh
GPU_NUM_DEVICES=8 python -m torch_xla.distributed.xla_spawn --num_gpus 8 run_mlm.py \
    --model_name_or_path bert-base-uncased \
    --dataset_name wikitext \
    --dataset_config_name wikitext-2-raw-v1 \
    --overwrite_output_dir true \
    --output_dir /tmp/test-mlm \
    --per_gpu_train_batch_size 16 \
    --do_eval \
    --fp16 true \
    --max_grad_norm 0 \
    --do_train \
    --num_train_epochs 3
```
Results of final training losses:
| Backend | fp16 | fp32       | fp16+max_grad_norm=0| fp32+max_grad_norm=0|
| --- | ----------- | --------| ----------- | --------| 
| pytorch cuda| 1.8649 | 1.8575 | 1.8753 | 1.8694 |
| torch_xla cuda| 1.86 | 1.8576 | 1.8694 | 1.867|

cc @sgugger

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1762,6 +1762,11 @@ def _inner_training_loop(
                 else:
                     tr_loss_step = self.training_step(model, inputs)
 
+                # Reduce gradients first for XLA device to be consistent with native pytorch behavior
+                if is_torch_tpu_available():
```
Comment (sgugger) [inline src/transformers/trainer.py:1766]: This should only be done if gradient scaling is necessary, otherwise it adds computation for nothing.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1819,7 +1820,7 @@ def _inner_training_loop(
                             self.scaler.step(self.optimizer)
                             self.scaler.update()
                         else:
-                            xm.optimizer_step(self.optimizer)
+                            self.optimizer.step()
```
Comment (sgugger) [inline src/transformers/trainer.py:1823]: This change should not be there.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this, added a few comments.

File: src/transformers/trainer.py (hunk 1 above)
Comment (ymwangg) [reply src/transformers/trainer.py:1766]: Could you elaborate more?
My understanding is `all_reduce` is necessary when running on multiple devices, regardless of if gradient scaling is used.
This PR essentially modified the sequence of operations from:
```py
loss = loss_fn(...)
loss.backward()
if max_grad_norm > 0:
    if do_grad_scaling > 0:
        xm.all_reduce(...)
        self.scaler.unscale_(self.optimizer)
    nn.utils.clip_grad_norm_(...)
if self.do_grad_scaling:
    self.scaler.step(self.optimizer)
    self.scaler.update()
else:
    xm.optimizer_step(self.optimizer)
```
to
```py
loss = loss_fn(...)
loss.backward()
xm.all_reduce(...)
if max_grad_norm > 0:
    if do_grad_scaling > 0:
        self.scaler.unscale_(self.optimizer)
    nn.utils.clip_grad_norm_(...)
if self.do_grad_scaling:
    self.scaler.step(self.optimizer)
    self.scaler.update()
else:
    self.optimizer.step()
```
Note `xm.optimizer_step() == xm.all_reduce(...) + optimizer.step()` [code](https://github.com/pytorch/xla/blob/28ea3758e9586ef8cc22270a16e1ddb4a21aa6f7/torch_xla/core/xla_model.py#L1021-L1022), here we need to decouple `xm.all_reduce` and `optimizer.step` in favor of the extra gradient clipping step in between.

For single device case, it won't add new computations in the hlo graph due to the flag [here](https://github.com/pytorch/xla/blob/28ea3758e9586ef8cc22270a16e1ddb4a21aa6f7/torch_xla/core/xla_model.py#L584), which evaluates to false for single device case.

File: src/transformers/trainer.py (hunk 2 above)
Comment (ymwangg) [reply src/transformers/trainer.py:1823]: As I explained in the previous comment, this is to decouple `xm.all_reduce` and `optimizer.step` in favor of the extra gradient clipping step in between.

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py:1766]: That's not the exact code that PyTorch XLA executes. I'd rather keep using the code from PyTorch XLA which is maintained than reproduce it here in all cases.
